NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/doc/clm.1
.\" Copyright (c) 2014 Stijn van Dongen
.TH "clm" 1 "16 May 2014" "clm 14-137" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
clm \- perform various computations on graphs and clusterings
.SH SYNOPSIS

.di ZV
.in 0
.nf \fC
   \fBclm\fP <mode> [mode-options] [mode-files]
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

.SH DESCRIPTION

\fBclm\fP implements a variety of computations on graphs and clusterings\&. The
first argument to \fBclm\fP should be a \fImode\fP, which is a string establishing
the type of computation to invoke\&. Each mode \fImode\fP is described in the
clm\fImode\fP manual page\&. The currently available modes are
\fBclm\ \&dist\fP,
\fBclm\ \&info\fP,
\fBclm\ \&meet\fP,
\fBclm\ \&mate\fP,
\fBclm\ \&close\fP,
\fBclm\ \&order\fP,
\fBclm\ \&imac\fP,
\fBclm\ \&vol\fP, and
\fBclm\ \&residue\fP\&.
The \fBdist\fP mode is thus described in the
\fBclmdist\fP manual page\&.

Invoking \fBclm\fP without arguments causes it to print out a list
of available modes with a short description of the type of
command line expected by that mode\&.

Invoking \fBclm\fP with just a mode will print out a longer listing
of options available for that particular mode\&.
Modes that are able to function normally without arguments
can be invoked by adding the \fI--nop\fP argument\&.

Several options are shared between all \fBclm\fP modes\&. Note that
these options are specified after the \fBmode\fP\&. See \fBOPTIONS\fP\&.

The \fB--version\fP option causes \fBclm\fP to print out version
and license information\&.

Issuing \fIclm help mode\fP will cause \fBclm\fP to look for a manual page
describing \fImode\fP and display it if found\&. This requires that the
MANPATH environment variable contains the directory in which the MCL-edge
manual pages were installed\&. This will usually be \fCPREFIX/share\fP,
where \fCPREFIX\fP is the path with which the software was configured\&.
.SH OPTIONS

These are options that pertain to all modes\&. They should be specified
\fIafter\fP the mode argument\&.

.ZI 2m "\fB-h\fP (\fIsynopsis\fP)"
\&
'in -2m
.ZI 2m "\fB--help\fP (\fIsynopsis\fP)"
\&
'in -2m
'in +2m
\&
.br
List available options\&.
.in -2m

.ZI 2m "\fB--nop\fP (\fIno-op\fP)"
\&
.br
Not an option\&. This option has no effect then to increment
the argument count\&. This can be useful for \fBclm\fP modes which are able to
function without any options\&. Such a mode typically reads from STDIN, writes
to STDOUT, and uses default settings\&. However, simply specifying a mode
without options leads \fBclm\fP to output a list of available options for that
mode\&. This can be averted by using the \fB--nop\fP option\&.
.in -2m

.ZI 2m "\fB-set\fP key=val (\fIset key-value pair\fP)"
\&
.br
Sets the key\ \&\fBkey\fP to value\ \&\fBval\fP in the environment\&.
Some modes allow adjustment of settings in this manner\&.
.in -2m

.ZI 2m "\fB-debug\fP <int> (\fIset debug level/flags\fP)"
\&
'in -2m
.ZI 2m "\fB--debug\fP (\fIturn default debugging on\fP)"
\&
'in -2m
.ZI 2m "\fB--test\fP (\fIturn default testing on\fP)"
\&
'in -2m
'in +2m
\&
.br
The first two turn on debugging, the last is for testing\&. The effect of
these options is otherwise undocumented\&.
.in -2m

.ZI 2m "\fB--progress\fP (\fIenable progress reporting\fP)"
\&
.br
This will enable some kind of progress reporting, by
writing to STDERR\&.
.in -2m
.SH AUTHOR

Stijn van Dongen\&.
.SH SEE ALSO

\fBmclfamily(7)\fP for an overview of all the documentation
and the utilities in the mcl family\&.


NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/src/shcl/clm.c
/*   (C) Copyright 1999, 2000, 2001, 2002, 2003, 2004 Stijn van Dongen
 *   (C) Copyright 2005, 2006, 2007 Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/


/* TODO
 *    mode-dispatch:
 *       check number of arguments left after exhaust.
 *       apropos should grep out appropriate options.
 *
 *    plugin mode where master C code does not need to be modified.
*/


#include <string.h>
#include <stdio.h>

#include "clm.h"
#include "clmmate.h"
#include "clmclose.h"
#include "clmorder.h"
#include "clmmeet.h"
#include "clmdist.h"
#include "clmimac.h"
#include "clmresidue.h"
#include "clmadjust.h"
#include "clmoptics.h"
#include "clminfo.h"
#include "clminfo2.h"
#include "clmps.h"

#include "impala/matrix.h"
#include "impala/io.h"
#include "impala/iface.h"
#include "impala/compose.h"
#include "impala/ivp.h"
#include "impala/app.h"

#include "clew/clm.h"

#include "util/types.h"
#include "util/err.h"
#include "util/opt.h"
#include "util/ding.h"


const char* clm_syntax =
   "Usage: clm <mode> [--nop] [options] [files]\n"
   "       clm <mode> -h (for mode specific options)"
   ;


enum
{  ID_MATE = 0
,  ID_MEET
,  ID_DIST
,  ID_IMAC
,  ID_INFO
,  ID_INFO2
,  ID_ORDER
,  ID_FRAME
,  ID_CLOSE
,  ID_VOL
,  ID_STABLE
,  ID_DAG
,  ID_FORMAT
,  ID_RESIDUE
,  ID_ADJUST
,  ID_OPTICS
,  ID_ENSTRICT
,  ID_FOLD
,  ID_PS
,  ID_UNUSED
}  ;


/* help function should get the name from the hook callback
 * function in clm_dir.
*/

mcxDispEntry clm_dir[] =
{  {  ID_MATE,    mcxDispHookMate  }
,  {  ID_MEET,    mcxDispHookMeet  }
,  {  ID_VOL,     mcxDispHookVol   }
,  {  ID_STABLE,  mcxDispHookStable}
,  {  ID_DIST,    mcxDispHookDist  }
,  {  ID_INFO,    mcxDispHookInfo  }
,  {  ID_INFO2,   mcxDispHookInfo2 }
,  {  ID_IMAC,    mcxDispHookImac  }
,  {  ID_ORDER,   mcxDispHookOrder }
,  {  ID_CLOSE,   mcxDispHookClose }
,  {  ID_RESIDUE, mcxDispHookResidue }
,  {  ID_ADJUST,  mcxDispHookAdjust }
,  {  ID_OPTICS,  mcxDispHookOptics }
,  {  ID_ENSTRICT, mcxDispHookEnstrict }
,  {  ID_FOLD,    mcxDispHookFold }
,  {  ID_PS,      mcxDispHookPS    }
,  {  -1,         NULL         }
}  ;


mcxOptAnchor clmSharedOptions[] =
{  {  "--version"
   ,  MCX_OPT_DEFAULT
   ,  CLM_DISP_VERSION
   ,  NULL
   ,  "output version information, exit"
   }
,  {  "--test"
   ,  MCX_OPT_DEFAULT
   ,  CLM_DISP_TEST
   ,  NULL
   ,  "test"
   }
,  {  "-progress"
   ,  MCX_OPT_HASARG
   ,  CLM_DISP_PROGRESS
   ,  "<int>"
   ,  "set progress interval size"
   }
,  {  "--progress"
   ,  MCX_OPT_DEFAULT
   ,  CLM_DISP_PROGRESS2
   ,  NULL
   ,  "turn on progress reporting"
   }
,  {  "-debug"
   ,  MCX_OPT_HASARG
   ,  CLM_DISP_DEBUG
   ,  "<int>"
   ,  "set debug level or bits"
   }
,  {  "--debug"
   ,  MCX_OPT_DEFAULT
   ,  CLM_DISP_DEBUG2
   ,  NULL
   ,  "turn on debugging"
   }
,  {  "-set"
   ,  MCX_OPT_HASARG
   ,  CLM_DISP_SET
   ,  "key=val"
   ,  "set key to val in ENV"
   }
,  {  "--nop"
   ,  MCX_OPT_DEFAULT
   ,  CLM_DISP_NOP
   ,  NULL
   ,  "this option has no affect but changing the argument count"
   }
,  {  "-h"
   ,  MCX_OPT_DEFAULT
   ,  CLM_DISP_HELP
   ,  NULL
   ,  "this"
   }
,  {  "--help"
   ,  MCX_OPT_DEFAULT
   ,  CLM_DISP_APROPOS
   ,  NULL
   ,  "this"
   }
,  {  "--amoixa"
   ,  MCX_OPT_HIDDEN
   ,  CLM_DISP_AMOIXA
   ,  NULL
   ,  "show hidden options too"
   }
,  {  NULL ,  0 ,  0 ,  NULL, NULL}
}  ;


mcxbits  mcx_debug_g     =  0;
unsigned mcx_progress_g  =  0;
mcxbool  mcx_test_g      =  FALSE;


mcxstatus sharedArgHandle
(  int optid
,  const char* val
,  mcxDispHook*   hook
,  mcxDispBundle* bundle
)
   {  mcxTing* full_syntax = mcxTingPrint(NULL, "%s %s", "clm", hook->syntax)
   ;  switch(optid)
      {  case CLM_DISP_HELP
      :  case CLM_DISP_APROPOS
      :  case CLM_DISP_AMOIXA
      :  
         mcxOptApropos
         (  stdout
         ,  hook->name
         ,  full_syntax->str
         ,  15
         ,  0
         ,  bundle->disp_shared
         )
      ;  mcxOptApropos
         (  stdout
         ,  hook->name
         ,  NULL
         ,  15
         ,     MCX_OPT_DISPLAY_SKIP
            |  (  optid == CLM_DISP_AMOIXA
               ?  MCX_OPT_DISPLAY_HIDDEN
               :  0
               )
         ,  hook->options
         )
      ;  mcxExit(0)
      ;  break
      ;

         case CLM_DISP_VERSION
      :  bundle->disp_version(hook->name)
      ;  mcxExit(0)
      ;  break
      ;

         case CLM_DISP_NOP
      :  NOTHING
      ;  break
      ;

         case CLM_DISP_TEST
      :  mcx_test_g = TRUE
      ;  break
      ;

         case CLM_DISP_PROGRESS
      :  mcx_progress_g = atoi(val)
      ;  break
      ;

         case CLM_DISP_PROGRESS2
      :  mcx_progress_g = 1
      ;  break
      ;

         case CLM_DISP_DEBUG
      :  mcx_debug_g = atoi(val)
      ;  break
      ;

         case CLM_DISP_DEBUG2
      :  mcx_debug_g = -1u
      ;  break
      ;

         case CLM_DISP_SET
      :  mcxSetenv(val)
      ;  break
      ;

         default
      :  break
      ;
      }
      mcxTingFree(&full_syntax)
   ;  return STATUS_OK
;  }


int main
(  int                  argc
,  const char*          argv[]
)
   {  mcxDispBundle bundle 

   ;  mcxLogLevel =
      MCX_LOG_AGGR | MCX_LOG_MODULE | MCX_LOG_GAUGE | MCX_LOG_WARN
   ;  mclxIOsetQMode("MCLXIOVERBOSITY", MCL_APP_VB_NO)
   ;  mclx_app_init(stderr)

   ;  bundle.disp_argc     =  argc
   ;  bundle.disp_argv     =  argv
   ;  bundle.disp_name     =  "clm"
   ;  bundle.disp_syntax   =  clm_syntax
   ;  bundle.disp_shared   =  clmSharedOptions
   ;  bundle.n_disp_shared =  sizeof(clmSharedOptions) / sizeof(mcxOptAnchor) -1;
   ;  bundle.disp_version  =  app_report_version
   ;  bundle.disp_table    =  clm_dir
   ;  bundle.shared_handler=  sharedArgHandle

   ;  return mcxDispatch(&bundle)
;  }


